plugins/module_utils/identity/bitwarden.py

```diff
@@ -0,0 +1,321 @@
+        self._folder = Folder(id_=value['folderId'])
+
+
+class Folder(_Element):
+    '''A folder in the password database.'''
+
+    organization = property(lambda self: self._organization)
+
+    def __init__(self, name=None, id_=None, organization=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify folder in multiple ways')
+
+        self._name = name
+        self._id = id_
+        self._organization = organization
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return folder dict where the specified attribute
+        has the specified value.
+        '''
+        matches = [folder for folder in self.organization.folders if folder.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Folder not found")
+        if len(matches) > 1:
+            raise AssertionError("Folder not unique")
+
+        return matches[0]
+
+    @property
+    def items(self):
+        query = ['list', 'items']
+
+        # Only restrict by folder if specified.
+        if self.name is not None:
+            # bw treats 'null' as items not belonging to any folder.
+            query.extend(['--folderid', self.id or 'null'])
+
+        # Filter by organization since `bw` ORs filters together.
+        return [Item(dict_=d) for d in Client.run_json(query) if self.organization.matches(d)]
+
+    def get_item(self, name=None, id_=None):
+        return Item(name, id_, self, self.organization)
+
+    def matches(self, folder):
+        '''Return whether folder matches a pattern indicated by self.
+
+        If self.name is None, any folder matches.
+        If self.name is '' or 'No Folder', only the null folder is matched.
+        All other folders are matched by id.
+        '''
+
+        if self.name is None:
+            return True
+
+        # No Folder is a special folder name in bw.
+        if self.name in ('', 'No Folder'):
+            return folder['folderId'] is None
+
+        return self.id == folder['folderId']
+
+
+class Organization(_Element):
+    '''An organization in the password database.'''
+
+    def __init__(self, name=None, id_=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify organization in multiple ways')
+
+        self._name = name
+        self._id = id_
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return organization dict where the specified attribute
+        has the specified value.
+        '''
+
+        matches = [org for org in organizations() if org.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Organization not found")
+        if len(matches) > 1:
+            raise AssertionError("Organization not unique")
```

Comment:
Hm. Is the reason this sounds off to you because of how we split up the modules and that I didn't document the function properly? If so, I can add some additional documentation to the function to make it seem less off :)

Previously, the code in the `module_utils.identity.bitwarden` module was in `modules.identity.bitwarden` (and `bitwarden_info` didn't exist). One of the preconditions listed in the module is that the organization name be unique.

In the first link you included, I believe the reason for the assert included here falls under "Validating the values in the argument to know whether it is valid." However, we cannot validate the input in this situation until we have done a lookup.

I also deliberately used `assert` (until the CI told me off!) since we could still return a legitimate value (the first response), even if the specific value returned was unspecified, if someone had opted to turn on Python optimizations.

Reply:
No, the reason it sounds off is because assertions were meant to be a form of debugging or parameter validation, not general-use exceptions.

Is it validation? The parameters (`attribute` and `value`) are not invalid in themselves, AFAIUI, but the fact that they bring more than one organization is not valid for *this* code. By validation I would assume things like `attribute is not None` or `value > 0`.